### ISSUE TITLE
Create Python bindings

### ISSUE THREAD

**Issue by mjovanc** [delta-rs/delta#29]
We should create a Python library with bindings to Delta. I suggest we create a directory in this repository and handle everything here for convenience. 

I propose we (at least):

- Create a directory called: `delta_py` and publish the library as `deltaml` on PyPi. 
- Set up a workflow for building Delta Rust libraries, build the Python and test against the Rust libraries.

**Comment by mjovanc** [comment]
We should perhaps use `pyo3` for this: https://github.com/PyO3/pyo3

**Comment by coddiw0mple** [comment]
I'd like to work on this

**Comment by mjovanc** [comment]
@coddiw0mple Could you fork this repo and perhaps come up with a proof-of-concept that we can take a look at?

**Comment by coddiw0mple** [comment]
@mjovanc Sure will work on one. I'm thinking of adding the bindings for the API needed for the MNIST example as a POC, should that be okay?

**Comment by coddiw0mple** [comment]
I plan to create a focused POC using the existing MNIST classification example. I'll use pyo3 to create Python bindings (only for the functions needed for this task) and package them up. Then I'll recreate the same example in Python using these bindings. This way, we can evaluate if this approach works well and check the performance.
Sound good? If so, I'll get started.

**Comment by mjovanc** [comment]
> I plan to create a focused POC using the existing MNIST classification example. I'll use pyo3 to create Python bindings (only for the functions needed for this task) and package them up. Then I'll recreate the same example in Python using these bindings. This way, we can evaluate if this approach works well and check the performance.
> 
> Sound good? If so, I'll get started.

Go ahead, we will evaluate. Don't work too much on this just a basic implementation so we can see if this is the right approach for us.

**Comment by mjovanc** [comment]
> Is the Python bindings something that is still being considered?

Yes, we will for sure have them. But at this stage it's not prioritized since we want to make the Rust APIs stable first for Classical ML before we think about that. But we are soon to release our MVP and then for Q4 we will have most algorithms in place and after that it could be relevant to start on that for those who can contribute to it.

**Comment by Rhaqim** [comment]
> > Is the Python bindings something that is still being considered?
> 
> Yes, we will for sure have them. But at this stage it's not prioritized since we want to make the Rust APIs stable first for Classical ML before we think about that. But we are soon to release our MVP and then for Q4 we will have most algorithms in place and after that it could be relevant to start on that for those who can contribute to it.

Okay that makes a lot of sense, I am still exploring all the other issues, I would love to contribute to this amazing project. Whenever the this becomes a priority i will also contribute to it.

**Comment by mjovanc** [comment]
> > > Is the Python bindings something that is still being considered?
> > 
> > 
> > Yes, we will for sure have them. But at this stage it's not prioritized since we want to make the Rust APIs stable first for Classical ML before we think about that. But we are soon to release our MVP and then for Q4 we will have most algorithms in place and after that it could be relevant to start on that for those who can contribute to it.
> 
> Okay that makes a lot of sense, I am still exploring all the other issues, I would love to contribute to this amazing project. Whenever the this becomes a priority i will also contribute to it.

Great. One thing that is really useful for the project at this stage before we make our MVP is to try it out and find bugs. Maybe mathematical/algorithmic issues that we didn't foreseen. 

If you could do that and submit PRs on bugs and stuff like that, that is extremely valuable for the project and good introduction to the code base to go through.

**Comment by Rhaqim** [comment]
> > > > Is the Python bindings something that is still being considered?
> > > 
> > > 
> > > Yes, we will for sure have them. But at this stage it's not prioritized since we want to make the Rust APIs stable first for Classical ML before we think about that. But we are soon to release our MVP and then for Q4 we will have most algorithms in place and after that it could be relevant to start on that for those who can contribute to it.
> > 
> > 
> > Okay that makes a lot of sense, I am still exploring all the other issues, I would love to contribute to this amazing project. Whenever the this becomes a priority i will also contribute to it.
> 
> Great. One thing that is really useful for the project at this stage before we make our MVP is to try it out and find bugs. Maybe mathematical/algorithmic issues that we didn't foreseen.
> 
> If you could do that and submit PRs on bugs and stuff like that, that is extremely valuable for the project and good introduction to the code base to go through.

Alright this is a clear enough path for me to start with, testing the classical algorithms, if i do find anything I will report them and create PRs.
